### PR TITLE
Censys:

### PR DESCRIPTION
**Pull request by dantavori**
## Status
Ready

## Related Issues
[https://github.com/demisto/etc/issues/18131](https://github.com/demisto/etc/issues/18131)

## Description
1. fixed bug described in issue #18131:
In the Censys API, when there are no results for a search, it throws a 404. Demisto is interpreting this as a playbook error when it's actually just the regular behavior of Censys.

2. Beautified the code a bit (no logic changes)

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Dependencies
Mention the dependencies of the entity you changed as given from the precommit hooks in checkboxes, and tick after tested them.
- [ ] Dependency 1
- [ ] Dependency 2
- [ ] Dependency 3

### CODE REVIEW

**Comment by Itay4** [comment]
@dantavori what are all the code changes and subtype removal for?
@yardensade fyi

**Comment by yardensade** [comment]
We need at least the insecure mandatory part cherry picked asap and merged into the release build

**Comment by dantavori** [comment]
@Itay4 just some cosmetic changes, nothing else (I've been told not to do this again before getting an official instruction to, but it was already done by then). Which subtype removal?

File: Integrations/Censys/Censys.yml [hunk 1]
```diff
@@ -76,7 +76,6 @@ script:
   runonce: false
   script: '-'
   type: python
-  subtype: python2
```
Comment (Itay4) [inline Integrations/Censys/Censys.yml]: this

**Comment by Itay4** [comment]
lets revert those changes, dont want to review and handle these right now
lets keep the only change to be the `required` from `true` to `false`

File: Integrations/Censys/Censys.yml (hunk 1 above)
Comment (dantavori) [reply Integrations/Censys/Censys.yml]: @ronykoz told me to add it

Comment (Itay4) [reply Integrations/Censys/Censys.yml]: add what? you removed it

**Comment by yardensade** [comment]
Let's make sure to merge this asap to master, this is really important. (at least the insecure part like in cherry pick PR -> https://github.com/demisto/content/pull/4164)

**Comment by Itay4** [comment]
@ronykoz @guykeller @dantavori we need to merge this asap

File: Integrations/Censys/Censys.py [hunk 2]
```diff
@@ -1,61 +1,70 @@
...
-            json.loads(data)
-        except ValueError:
-            return '### Error: Data is not in JSON format'
+def test_module():
+    url_suffix = "view/ipv4/8.8.8.8"
+    res = requests.get(API_URL + url_suffix, auth=(API_ID, API_SECRET), verify=USE_SSL)
+    if res.status_code == 200:
+        demisto.results('ok')
+    else:
+        demisto.results('test failed')
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: return the real response, this info will not give the user an informative enough problem

File: Integrations/Censys/Censys.py [hunk 3]
```diff
@@ -1,61 +1,70 @@
...
+
 
-    res = requests.request(method, url + uri, auth=(user, passwd), data=data, verify=USE_SSL)
+def send_request(method, url_suffix, data=None):
+    res = requests.request(method, API_URL + url_suffix, auth=(API_ID, API_SECRET),
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: separate into several lines

File: Integrations/Censys/Censys.py [hunk 4]
```diff
@@ -1,61 +1,70 @@
...
+    }
+    raw = send_request('POST', url_suffix, data)
+    readable = tableToMarkdown("Search results for {0} in {1}".format(query, index), raw["results"])
+    return_outputs(readable, raw)
+
+
+''' EXECUTION CODE '''
+command = demisto.command()
+LOG('command is {0}'.format(command))
+try:
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: add handle_proxy() and proxy to the yml file

File: Integrations/Censys/Censys.py [hunk 5]
```diff
@@ -1,61 +1,70 @@
...
 
-    if res.status_code >= 400:
-        if res.text.startswith('{'):
-            return '### Error: ' + str(res.status_code) + ': ' + res.json()["error"]
-        else:
-            return '### Error: ' + str(res.status_code)
+    if res.status_code == 404:
+        return None
+    elif res.status_code >= 400:
+        return_error("Error {0}: {1}".format(res.status_code, data["error"].title()))
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]:
```suggestion
        return_error("Received and error - status code [{0}], error message: {1}".format(res.status_code, data["error"].title()))
```

File: Integrations/Censys/Censys.py [hunk 6]
```diff
@@ -1,61 +1,70 @@
...
+
+''' EXECUTION CODE '''
+command = demisto.command()
+LOG('command is {0}'.format(command))
+try:
+    args = demisto.args()
+    if command == 'test-module':
+        test_module()
+    elif command == 'cen-view':
+        censys_view_command(**args)
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: dont use **args, fix this as we talked about

File: Integrations/Censys/Censys.py [hunk 7]
```diff
@@ -1,61 +1,70 @@
...
+    elif command == 'cen-search':
+        censys_search_command(**args)
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: same here

**Comment by ronykoz** [comment]
@Itay4 we will merge this today

File: Integrations/Censys/Censys.py (hunk 4 above)
Comment (dantavori) [reply Integrations/Censys/Censys.py]: Please check if I've made the changes correctly (in general I'd love to understand more about this option and about fetch incidents, is there any documentation for them?)

File: Integrations/Censys/CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,6 @@
 ## [Unreleased]
-
+  - Bug fix
```
Comment (ronykoz) [inline Integrations/Censys/CHANGELOG.md]: Which bug?

File: Integrations/Censys/CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,6 @@
...
+  - Proxy parameter was missing
```
Comment (ronykoz) [inline Integrations/Censys/CHANGELOG.md]:
```suggestion
  - Added proxy support
```

File: Integrations/Censys/Censys.py [hunk 10]
```diff
@@ -1,61 +1,80 @@
...
-            json.loads(data)
-        except ValueError:
-            return '### Error: Data is not in JSON format'
+def test_module():
+    url_suffix = "view/ipv4/8.8.8.8"
+    res = send_request('GET', url_suffix)
+    if res is not None:
+        demisto.results('ok')
+    else:
+        return_error('Error 404: Test failed.')
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: Are you sure it's 404?

File: Integrations/Censys/Censys.yml [hunk 11]
```diff
@@ -20,6 +20,10 @@ configuration:
   name: insecure
   required: false
   type: 8
+- display: Use system proxy
```
Comment (ronykoz) [inline Integrations/Censys/Censys.yml]:
```suggestion
- display: Use system proxy settings
```

File: Integrations/Censys/Censys.py [hunk 12]
```diff
@@ -1,61 +1,80 @@
...
-    if res.status_code >= 400:
-        if res.text.startswith('{'):
-            return '### Error: ' + str(res.status_code) + ': ' + res.json()["error"]
-        else:
-            return '### Error: ' + str(res.status_code)
+def send_request(method, url_suffix, data=None):
+    res = requests.request(method, API_URL + url_suffix,
+                           auth=(API_ID, API_SECRET),
+                           data=data, verify=USE_SSL)
+    data = json.loads(res.text)
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: better to check that res is not empty

File: Integrations/Censys/Censys.py [hunk 13]
```diff
@@ -1,61 +1,81 @@
...
-            json.loads(data)
-        except ValueError:
-            return '### Error: Data is not in JSON format'
+def test_module():
+    url_suffix = "view/ipv4/8.8.8.8"
+    res = send_request('GET', url_suffix)
+    if res is not None:
+        demisto.results('ok')
+    else:
+        return_error("Received an error, data retrieved: {1}".format(res))
```
Comment (ronykoz) [inline Integrations/Censys/Censys.py]: better to use contents/text

**Comment by ronykoz** [comment]
@kirbles19 please review the following file:
https://github.com/demisto/content/edit/Censys-bug-fix/Integrations/Censys/CHANGELOG.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F4112

**Comment by kirbles19** [comment]
@dantavori @ronykoz Edited .yml and .md. All done.

**Pull request merged**